OpenRA.Game/Traits/World/MusicPlaylist.cs

```diff
@@ -0,0 +1,150 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.GameRules;
+
+namespace OpenRA.Traits
+{
+	[Desc("Trait for music handling. Attach this to the world actor.")]
+	public class MusicPlaylistInfo : ITraitInfo
+	{
+		public readonly string StartingMusic = null;
+		public readonly bool LoopStartingMusic = false;
+
+		public object Create(ActorInitializer init) { return new MusicPlaylist(init.World, this); }
+	}
+
+	public class MusicPlaylist : INotifyActorDisposing
+	{
+		readonly MusicInfo[] random;
+		readonly MusicInfo[] playlist;
+
+		MusicInfo currentSong;
+		bool isMusicInstalled;
+		bool repeat;
+
+		public MusicPlaylist(World world, MusicPlaylistInfo info)
+		{
+			isMusicInstalled = world.Map.Rules.InstalledMusic.Any();
+
+			playlist = world.Map.Rules.InstalledMusic.Select(a => a.Value).ToArray();
+			if (!isMusicInstalled)
+				return;
+
+			random = playlist.Shuffle(Game.CosmeticRandom).ToArray();
+
+			if (Game.Settings.Sound.MapMusic && !string.IsNullOrEmpty(info.StartingMusic) && world.Map.Rules.Music.ContainsKey(info.StartingMusic))
+			{
+				currentSong = world.Map.Rules.Music[info.StartingMusic];
+				repeat = info.LoopStartingMusic;
+			}
+			else
+			{
+				currentSong = Game.Settings.Sound.Shuffle ? random.First() : playlist.First();
+				repeat = Game.Settings.Sound.Repeat;
+			}
+
+			Play();
+		}
+
+		public MusicInfo CurrentSong()
+		{
+			return currentSong;
+		}
+
+		public MusicInfo[] AvailablePlaylist()
+		{
+			return playlist;
+		}
+
+		public bool IsInstalled()
+		{
+			return isMusicInstalled;
+		}
+
+		void Play()
+		{
+			if (currentSong == null)
+				return;
+
+			Sound.PlayMusicThen(currentSong, () =>
+			{
+				if (!isMusicInstalled)
+					return;
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
+			});
+		}
+
+		public void Play(MusicInfo music)
+		{
+			if (music == null)
+				return;
+
+			currentSong = music;
+			repeat = Game.Settings.Sound.Repeat;
+
+			Sound.PlayMusicThen(music, () =>
+			{
+				if (!isMusicInstalled)
+					return;
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
+			});
+		}
+
+		public void Play(MusicInfo music, Action onComplete)
+		{
+			if (music == null)
+				return;
+
+			currentSong = music;
+			Sound.PlayMusicThen(music, onComplete);
+		}
+
+		public MusicInfo GetNextSong()
+		{
+			if (!isMusicInstalled)
+				return null;
+
+			var songs = Game.Settings.Sound.Shuffle ? random : playlist;
+			return songs.SkipWhile(m => m != currentSong)
+				.Skip(1).FirstOrDefault() ?? songs.FirstOrDefault();
+		}
+
+		public MusicInfo GetPrevSong()
```

Comment:
I got confused in the syntax, so I just made the two public functions into wrappers of a private one instead.